File: tests/test_class_sh_trampoline_basic.cpp
#include "pybind11_tests.h"

#include <memory>

namespace pybind11_tests {
namespace class_sh_trampoline_basic {

struct Abase {
    int val = 0;
    virtual ~Abase() = default;
    explicit Abase(int val_) : val{val_} {}
    int Get() const { return val * 10 + 3; }
    virtual int Add(int other_val) const = 0;

    // Some compilers complain about implicitly defined versions of some of the following:
    Abase(const Abase &) = default;
    Abase(Abase &&) noexcept = default;
    Abase &operator=(const Abase &) = default;
    Abase &operator=(Abase &&) noexcept = default;
};

struct AbaseAlias : Abase, py::trampoline_self_life_support {
    using Abase::Abase;

    int Add(int other_val) const override {
        PYBIND11_OVERRIDE_PURE(int,   /* Return type */
                               Abase, /* Parent class */
                               Add,   /* Name of function in C++ (must match Python name) */
                               other_val);
    }
};

int AddInCppRawPtr(const Abase *obj, int other_val) { return obj->Add(other_val) * 10 + 7; }

int AddInCppSharedPtr(const std::shared_ptr<Abase> &obj, int other_val) {
    return obj->Add(other_val) * 100 + 11;
}

int AddInCppUniquePtr(std::unique_ptr<Abase> obj, int other_val) {
    return obj->Add(other_val) * 100 + 13;
}

} // namespace class_sh_trampoline_basic
} // namespace pybind11_tests

using namespace pybind11_tests::class_sh_trampoline_basic;

TEST_SUBMODULE(class_sh_trampoline_basic, m) {
    py::classh<Abase, AbaseAlias>(m, "Abase")
        .def(py::init<int>(), py::arg("val"))
        .def("Get", &Abase::Get)
        .def("Add", &Abase::Add, py::arg("other_val"));

    m.def("AddInCppRawPtr", AddInCppRawPtr, py::arg("obj"), py::arg("other_val"));
    m.def("AddInCppSharedPtr", AddInCppSharedPtr, py::arg("obj"), py::arg("other_val"));
    m.def("AddInCppUniquePtr", AddInCppUniquePtr, py::arg("obj"), py::arg("other_val"));
}

File: tests/test_class_sh_trampoline_basic.py
from __future__ import annotations

from pybind11_tests import class_sh_trampoline_basic as m


class PyDrvd(m.Abase):
    def __init__(self, val):
        super().__init__(val)

    def Add(self, other_val):
        return self.Get() * 100 + other_val


def test_drvd_add():
    drvd = PyDrvd(74)
    assert drvd.Add(38) == (74 * 10 + 3) * 100 + 38


def test_drvd_add_in_cpp_raw_ptr():
    drvd = PyDrvd(52)
    assert m.AddInCppRawPtr(drvd, 27) == ((52 * 10 + 3) * 100 + 27) * 10 + 7


def test_drvd_add_in_cpp_shared_ptr():
    while True:
        drvd = PyDrvd(36)
        assert m.AddInCppSharedPtr(drvd, 56) == ((36 * 10 + 3) * 100 + 56) * 100 + 11
        return  # Comment out for manual leak checking (use `top` command).


def test_drvd_add_in_cpp_unique_ptr():
    while True:
        drvd = PyDrvd(25)
        assert m.AddInCppUniquePtr(drvd, 83) == ((25 * 10 + 3) * 100 + 83) * 100 + 13
        return  # Comment out for manual leak checking (use `top` command).
